### PR TITLE
use node-pre-gyp to pre-build apm-bindings

### PR DESCRIPTION
**Pull request by bmacnaughton**
### notable items
- appoptics-bindings => @appoptics/apm-bindings
- extend some test timeouts/ranges for slow gh-actions
- use package-lock.json again
- restructure github actions for multiple actions
  - the action chooses the script to run

### details
This will be published to the `@appoptics` namespace under the name `apm-bindings` as it is a major change. It will maintain the version sequencing of `appoptics-bindings`, the non-namespaced predecessor. Because this is a new name and a scoped namespace, a new version of the agent will be required in order to use it. That release can be either namespaced or not (or there can be both a namespaced and non-namespaced version that use this release).

The biggest change is that this will use [node-pre-gyp](https://github.com/mapbox/node-pre-gyp/). Most of the changes in this PR are to modify the github actions. github actions were previously used only for testing; now there is a publish action as well. It must be manually triggered. One item that must be resolved before merging this PR is to decide the name of the bucket that node-pre-gyp will target for the production binaries as well as appropriate properties for that bucket (versioning, object-lock, etc.). The bucket must be public or it is not possible for node-pre-gyp to read from it. node-pre-gyp adds '--acl public-read' to the s3 put-object api call.

~~At this moment the node-pre-gyp dependency is tied to a specific github commit as features required by our release are not in the existing node-pre-gyp codebase.~~

the `napi_versions` array in the `binary` property of package.json is set for the [lowest version we want to support](https://nodejs.org/api/n-api.html#n_api_n_api_version_matrix). it's possible to add additional versions but, while there might be minor improvements, the code doesn't use features available in higher versions, so it just seems cleaner to keep things simple.

Testing the publishing is done by manually executing the `publish using node-pre-gyp for pre-built glibc and musl versions` workflow. you must enter two inputs: the branch/tag and whether to publish to production or not. at the time of this comment production is just another bucket with no special characteristics.

manual tests that have been performed:
- no public access for bucket (publishing fails)
- public access empty bucket (publishing succeeds)
- public access but version already published (publishing fails)

### open items
- decide the name of the bucket that will be used for the production binaries as well as appropriate properties for that bucket (versioning, object-lock, etc.)
- running the publish script with the staging target does not automatically delete an existing target with the same spec. if the same version has been published the publish script will fail; in order to verify that publishing is working correctly use the s3 console or s3 cli to delete the conflicting targets.
- ~~see if node-pre-gyp can be released (or just keep commit hash and github install)~~ have an rc now

### CODE REVIEW

**Comment by bmacnaughton** [comment]
this visual overview of how the github actions work might be helpful
github-actions

**Comment by bmacnaughton** [comment]
i've added some minor cleanup - extended the test timeouts when running in a CI environment and moving my global eslintrc file into the repo.

File: .github/actions/generic/image-actions/common.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
+#!/bin/sh -l
+
+#
+# this file looks at the first argument to determine whether to run the
+# node-os-tests.sh script or the publish-bindings.sh script.
```
Comment (prettycoder) [inline .github/actions/generic/image-actions/common.sh:5]: has this be renamed from publish-bindings.sh to publish.sh

File: .github/actions/generic/image-actions/common.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        echo "args: $*"
+        exit 1
+    fi
+elif [ "$script_name" = "publish.sh" ]; then
+    if [ $# -ne 8 ]; then
+        echo "? error: publish.sh requires 8 arguments, not $#"
+        echo "args: $*"
+        exit 1
+    fi
+else
```
Comment (prettycoder) [inline .github/actions/generic/image-actions/common.sh:29]: not sure if I'm missing something but I can only find 4 resp. 7 args in those scripts. I don't think $0 counts.

**Review by prettycoder — COMMENTED** [review]
just 2 things I noticed while previewing
I'm also wondering about ubuntu/debian, there is no Dockerfile for one of those

File: .github/actions/generic/image-actions/common.sh (hunk 2 above)
Comment (bmacnaughton) [reply .github/actions/generic/image-actions/common.sh:29]: got this during our conversation.

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        # specify the branch/tag that to be published
+        branch: ${{github.event.inputs.tag-to-publish}}
+        # repurposed for publishing. (for testing it is a token to enable connecting
+        # to the production collector; for publishing it provides access to a potentially
+        # private repo.)
+        # TODO - the secret should become an organizational secret so it works for any account
+        # this one is tied to bmacnaughton (even though i use github.actor). not sure there is
+        # any other way to handle it unless ${{secrets.REPO_READ_${{github.actor}}}} syntax would
+        # work. i haven't tried it.
+        ao-token-prod: "https://${{github.actor}}:${{secrets.PRIVATE_REPO_READ}}@github.com/${{github.repository}}.git"
```
Comment (cheempz) [inline .github/workflows/publish.yml:53]: This wouldn't be needed in the actual public `appoptics-bindings-node` repo, right?

**Review by cheempz — APPROVED** [review]
LGTM, thanks @bmacnaughton.  I appreciate all the hoops you had to jump through to get this working, but after playing around with actions in a test repo https://github.com/cheempz/test yesterday it seems possible to achieve matrix testing using docker images pulled from dockerhub, so long term I'd like to migrate over to that instead of writing our own specific docker action.  

I do agree at this stage it's more important to proceed with existing PR and test out how things work with all the changes!

File: .github/workflows/publish.yml (hunk 3 above)
Comment (bmacnaughton) [reply .github/workflows/publish.yml:53]: no, that's required only for checking out of a private repo. there's not harm in keeping it there for now as it allows a place for experimentation without muddying the public repo.

**Comment by bmacnaughton** [comment]
if you want to pull from docker hub you can. the negative for that is that the dockerfiles for those images are disconnected from the process. right now pretty much everything you need to know about testing/publishing is here.

it's possible to keep the dockerfiles here and create a script to update them when changes are needed but is just another step that needs to be coordinated. my two bits.

**Pull request merged**